src/Notepads/Views/NotepadsMainPage.IO.cs

```diff
@@ -120,9 +120,7 @@ private async Task<bool> Save(ITextEditor textEditor, bool saveAs, bool ignoreUn
             StorageFile file = null;
             try
             {
-                if (textEditor.EditingFile == null || saveAs ||
-                    FileSystemUtility.IsFileReadOnly(textEditor.EditingFile) ||
```

Comment:
FileSystemUtility.IsFileReadOnly(textEditor.EditingFile)
You need to keep this because for files those are actually read-only, it should prompt save picker.

Reply:
Another question for you is that what if the file been dragged into notepads is actually a read-only file?

Reply:
I did it because as you said, when dragging a file into Notepads, it automatically becomes a "fake" Read Only, so it would not enter the save file function, where it has the trick to check if the file is really Ready Only.

When a file is a real Read Only and the user tries to save it, Notepads automatically displays a message stating that it cannot be saved to it. I believe that this way it becomes more explanatory for the user to know the reason for not being able to overwrite.
![Capture](https://user-images.githubusercontent.com/1365683/81366183-6290d180-90c0-11ea-9f10-fd07088ce14c.JPG)

Reply:
Not true since:

1. This is not obvious to user, since this is an error popup.
2. People don't read exception message.
3. "Access is denied" only makes sense to developers not general audience. I am pretty sure some of our users will not map it to "File is read only, please do save as"

That is also why I am popping up save dialog when this happens.

Reply:
Okay. I will implement the prompt Save As when save fails.

Reply:
> Okay. I will implement the prompt Save As when save fails.

I personally against using Path.IO since this is a kind of "expliot" to me.

But if you really want to do it, here is what I would suggest:

1. Whenever file is been dragged into notepads, you somehow mark it as "OpenedByDragging".

2. You check this flag explicitly when Save triggers. If you don't see this flag. Just follow the exciting code path and don't touch it.

3. If you are seeing this flag. Do below steps:
    1) Check if file is writable, if it is, then we are happy and you can do Path.IO.
    2) If it is not writable, fallback to open the save picker. (This is the case where the file was read-only before dragging into notepads)

Reply:
@JasonStein 

I was carrying out the implementation according to your guidelines, but in the course of it, I found that there is no way to know if the file is really writable without trying to write to it.

I believe that the best way to do this is to control exceptions. For example, if you are unable to write, throw an exception to open File Save Picker. What do you think?

Reply:
I don't quite like that approach tho (also exception message varies between different locales), let me take a look first. Also, what about the case when you drag a file into Notepads and then exit the app and reopen and then do save?

Reply:
I'm reopening the pull request for you to look at the changes.

If you look at the commits tree, you will notice that I was implementing "isFileOpenedByDragging", but as there was no way to check if the file is a real read only without writing to it, I chose to abandon it and try the exception path.

Watch this demo video
[![](https://img.youtube.com/vi/mppWY_1wGRA/hqdefault.jpg)](https://youtu.be/mppWY_1wGRA)
https://youtu.be/mppWY_1wGRA

Reply:
I tested it here and it worked. Good job Maickonn

Reply:
@Maickonn  You can remove all the changes related to "isFileOpenedByDragging'. I will test other scenarios in the meantime.

Reply:
> @Maickonn You can remove all the changes related to "isFileOpenedByDragging'. I will test other scenarios in the meantime.

Okay.

Reply:
@JasonStein it's done.

Reply:
On top of what you have tested, there are few more cases you need to test as well:

1. Open file from SMB share and disconnect, then try to save.
2. Open file from CloudDrive like OneDrive or Dropbox, disconnect it and then try to save.
3. Open file from a USB drive, disconnect it and then try to save.
4. Make sure session snapshot is enabled then you open the file, do some changes, close the app and reopen the app. Now try to save.

We need to make sure all above cases work as expected and show proper message to the user accordingly.

Reply:
> On top of what you have tested, there are few more cases you need to test as well:
> 
> 1. Open file from SMB share and disconnect, then try to save.
> 2. Open file from CloudDrive like OneDrive or Dropbox, disconnect it and then try to save.
> 3. Open file from a USB drive, disconnect it and then try to save.
> 4. Make sure session snapshot is enabled then you open the file, do some changes, close the app and reopen the app. Now try to save.
> 
> We need to make sure all above cases work as expected and show proper message to the user accordingly.

As requested, I tested on all listed scenarios. When the file cannot be written or is no longer accessible, Notepads opens a Save As.

About session snapshot, it is working correctly, even when the file has been opened by drag and drop.